### PR TITLE
Add a WEBIRC module

### PR DESCRIPTION
**Pull request by Heufneutje**
This should allow better control over people connecting from places like Mibbit and Kiwi. It's just the basic implementation of the WEBIRC command, not a full CGI:IRC implementation. I wanted to do a full implementation, but could not find any good documentation on the other mechanisms, nor could I find any services that still use anything other than WEBIRC.

### CODE REVIEW

**Comment by ElementalAlchemist** [comment]
WEBIRC should be sufficient since, as far as I know, everything's standardized on that, now. (I also don't know that there's ever been documentation on CGI:IRC.)

I'll have a look at this now.

File: txircd/modules/extra/webirc.py [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+					raise ConfigValidationError("webirc_allowed_sources", "password value must be a string")
+
+	def parseParams(self, user, params, prefix, tags):
+		if user.isRegistered():
+			user.sendSingleError("WebircCmd", irc.ERR_ALREADYREGISTERED, "WEBIRC may only be used during registration")
+			return None
+		if len(params) < 4:
+			user.sendSingleError("WebircCmd", irc.ERR_NEEDMOREPARAMS, "WEBIRC", "Not enough parameters")
+			return None
+		if user.ip not in self.ircd.config.get("webirc_allowed_sources", {}) and user.realHost not in self.ircd.config.get("webirc_allowed_sources", {}):
```
Comment (ElementalAlchemist) [inline txircd/modules/extra/webirc.py]: Should this be in a commandpermission action?

File: txircd/modules/extra/webirc.py [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+			self.ircd.log.warn("WEBIRC was requested from IP \"{ip}\", but the IP does not match any WEBIRC configuration.", ip=user.ip)
+			return None
+		if self.ircd.config["webirc_allowed_sources"][user.ip] != params[0]:
```
Comment (ElementalAlchemist) [inline txircd/modules/extra/webirc.py]: Should this also be in a commandpermission action?

File: txircd/modules/extra/webirc.py [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+			self.ircd.log.warn("WEBIRC was requested from IP \"{ip}\" with password \"{password}\", but this password does not match the WEBIRC configuration for this IP.", ip=user.ip, password=params[0])
+			return None
```
Comment (ElementalAlchemist) [inline txircd/modules/extra/webirc.py]: I'm not familiar with the WEBIRC spec or implementations of it, but should there be some sort of feedback on a failed password, or no?

I'm totally fine with not, but I want to make sure this is expected.

File: txircd/user.py [hunk 4]
```diff
@@ -844,6 +835,17 @@ def modeString(self, toUser):
 			return "{} {}".format("".join(modeStr), " ".join(params))
 		return "".join(modeStr)
 
+	def resolveDNS(self, ip):
```
Comment (ElementalAlchemist) [inline txircd/user.py]: This doesn't really have much to do with the user. Should it be in utils?

If nothing else, ircd might even be a better place for it than user.

File: txircd/modules/extra/webirc.py [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+		# We don't need params[1]; this is the client name which we don't use. We verify that the DNS resolution is correct and set the provided IP as the host if it is incorrect.
+		host = params[2]
+		ip = params[3]
+		if not isValidHost(host) or len(host) > self.ircd.config.get("hostname_length", 64) or user.resolveDNS(params[3]) != host:
+			self.ircd.log.warn("DNS resolution for WEBIRC command from IP \"{ip}\" with requested IP \"{requestip}\" and requested host \"{requesthost}\" has failed; using the requested IP address as the host instead.", ip=user.ip, requestip=ip, requesthost=host)
```
Comment (ElementalAlchemist) [inline txircd/modules/extra/webirc.py]: For logging, you should use `{user.ip}` in the string and `user=user` as a parameter. Same applies to all of the other log calls above and below.

File: txircd/modules/extra/webirc.py [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+			host = ip
+		return {
+			"ip": ip,
+			"host": host
+		}
+
+	def execute(self, user, data):
+		self.ircd.log.info("WEBIRC detected for for IP \"{ip}\"; changing their IP to \"{requestip}\" and their real host to \"{requesthost}\".", ip=user.ip, requestip=data["ip"], requesthost=data["host"])
```
Comment (ElementalAlchemist) [inline txircd/modules/extra/webirc.py]: "for for"

File: txircd/modules/extra/webirc.py [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+		ip = params[3]
+		if not isValidHost(host) or len(host) > self.ircd.config.get("hostname_length", 64) or user.resolveDNS(params[3]) != host:
+			self.ircd.log.warn("DNS resolution for WEBIRC command from IP \"{ip}\" with requested IP \"{requestip}\" and requested host \"{requesthost}\" has failed; using the requested IP address as the host instead.", ip=user.ip, requestip=ip, requesthost=host)
+			host = ip
+		return {
+			"ip": ip,
+			"host": host
+		}
+
+	def execute(self, user, data):
```
Comment (ElementalAlchemist) [inline txircd/modules/extra/webirc.py]: Command handlers should return `True` to indicate a successful handling.

File: txircd/user.py (hunk 4 above)
Comment (Heufneutje) [inline txircd/user.py]: I thought of that after I submitted the PR and you are most definitely right. Will fix later.

File: txircd/modules/extra/webirc.py (hunk 1 above)
Comment (Heufneutje) [inline txircd/modules/extra/webirc.py]: Will fix later.

File: txircd/modules/extra/webirc.py (hunk 2 above)
Comment (Heufneutje) [inline txircd/modules/extra/webirc.py]: Will fix later.

File: txircd/modules/extra/webirc.py (hunk 5 above)
Comment (Heufneutje) [inline txircd/modules/extra/webirc.py]: Fixed.

**Comment by Heufneutje** [comment]
Fixed a bunch of things. I'll fix everything else up later.

**Comment by Heufneutje** [comment]
Alright. I've addressed all of your comments now.

File: txircd/modules/extra/webirc.py [hunk 8]
```diff
@@ -0,0 +1,63 @@
+from twisted.plugin import IPlugin
+from twisted.words.protocols import irc
+from txircd.config import ConfigValidationError
+from txircd.module_interface import Command, ICommand, IModuleData, ModuleData
+from txircd.utils import isValidHost, resolveHost
+from zope.interface import implements
+
+irc.ERR_ALREADYREGISTERED = "462"
```
Comment (ElementalAlchemist) [inline txircd/modules/extra/webirc.py]: Get this outta here :P

File: txircd/modules/extra/webirc.py [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+		if "webirc_allowed_sources" in config:
+			if not isinstance(config["webirc_allowed_sources"], dict):
+				raise ConfigValidationError("webirc_allowed_sources", "value must be a dictionary")
+			for ip, password in config["webirc_allowed_sources"].iteritems():
+				if not isinstance(ip, basestring):
+					raise ConfigValidationError("webirc_allowed_sources", "ip value must be a string")
+				if not isinstance(password, basestring):
+					raise ConfigValidationError("webirc_allowed_sources", "password value must be a string")
+
+	def checkSourceAndPass(self, user, data):
```
Comment (ElementalAlchemist) [inline txircd/modules/extra/webirc.py]: You missed a return at the end of the function.

(Explicitly `return None` to make it obvious you intended that and didn't forget a different return value.)

**Comment by Heufneutje** [comment]
Alright, all fixed.

**Pull request merged**